soa-derive-internal/src/vec.rs

```diff
@@ -71,97 +88,144 @@ pub fn derive(input: &Input) -> TokenStream {
             /// reserving the same `additional` space for all fields.
             pub fn reserve(&mut self, additional: usize) {
-                #(self.#fields_names_1.reserve(additional);)*
+                #(self.data.#fields_names_1.reserve(self.len, additional);)*
             }
 
             /// Similar to [`
             #[doc = #vec_name_str]
             /// ::reserve_exact()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.reserve_exact)
             /// reserving the same `additional` space for all fields.
             pub fn reserve_exact(&mut self, additional: usize) {
-                #(self.#fields_names_1.reserve_exact(additional);)*
-            }
-
-            /// Similar to [`
-            #[doc = #vec_name_str]
-            /// ::shrink_to_fit()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.shrink_to_fit)
-            /// shrinking all fields.
-            pub fn shrink_to_fit(&mut self) {
-                #(self.#fields_names_1.shrink_to_fit();)*
+                #(self.data.#fields_names_1.reserve_exact(self.len, additional);)*
             }
 
             /// Similar to [`
             #[doc = #vec_name_str]
             /// ::truncate()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.truncate)
             /// truncating all fields.
             pub fn truncate(&mut self, len: usize) {
-                #(self.#fields_names_1.truncate(len);)*
+                unsafe {
+                    // Destroy the elements that are outside the given length
+                    while len < self.len {
+                        let i = self.len - 1;
+
+                        // Decrement len before calling drop_in_place() so a panic on Drop
+                        //   doesn't try to drop it a second time.
+                        self.len -= 1;
+
+                        #(ptr::drop_in_place(self.data.#fields_names_1.ptr().add(i));)*
+                    }
+                }
             }
 
             /// Similar to [`
             #[doc = #vec_name_str]
             /// ::push()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.push).
             pub fn push(&mut self, value: #name) {
+                fn write_to_raw_vec<T>(buf: &mut RawVec<T>, value: T, index: usize) {
+                    unsafe {
+                        let ptr = buf.ptr().add(index);
+                        ptr::write(ptr, value);
+                    }
+                }
+
                 let #name{#(#fields_names_1),*} = value;
-                #(self.#fields_names_1.push(#fields_names_2);)*
+                self.reserve(1);
+                #(write_to_raw_vec(&mut self.data.#fields_names_1, #fields_names_2, self.len);)*
+                self.len += 1;
             }
 
             /// Similar to [`
             #[doc = #vec_name_str]
             /// ::len()`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.len),
             /// all the fields should have the same length.
```

Comment:
this part of the doc is no longer required \o/

Reply:
:D